File: Android/Skola/Kanban/app/src/main/java/com/example/kanban/TaskDB.java
package com.example.kanban;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;
import java.util.Date;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Database(entities = {Task.class}, version = 1, exportSchema = false)
public abstract class TaskDB extends RoomDatabase {

    public abstract TaskDao getDao();
    private static volatile TaskDB INSTANCE;
    private static final int NUMBER_OF_THREADS = 4;
    static final ExecutorService databaseWriteExecutor = Executors.newFixedThreadPool(NUMBER_OF_THREADS);

    /**
     * @param context The context of the application for the database
     * @return A new instance of the TaskDB database
     */
    static TaskDB getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (TaskDB.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                                    TaskDB.class, "taskDatabase")
                            .addCallback(TaskDatabaseCallback)
                            .build();
                }
            }
        }
        return INSTANCE;
    }

    /**
     * Override the onCreate method to populate the database.
     * We clear the database every time it is created.
     */
    private static final RoomDatabase.Callback TaskDatabaseCallback = new RoomDatabase.Callback() {
        @Override
        public void onCreate(@NonNull SupportSQLiteDatabase db) {
            super.onCreate(db);
            /* This method is used to interact with the TaskDao class and by extension, the database */
            databaseWriteExecutor.execute(() -> {
                // Populate the database in the background.
                // If you want to start with more Tasks, just add them.
                TaskDao dao = INSTANCE.getDao();
                // Clears the database table at creation (or app install)
                dao.clearTable();
                Date timeNow = new Date();
                Task welcome = new Task(
                    "Welcome to Kanban!",
                    "Create a new Task by clicking the '+' button down below in the corner of your screen",
                    ViewHolder.NOT_STARTED,timeNow.toString()
                );
                dao.insert(welcome);
            });
        }
    };
}

File: Android/Skola/KanbanKt/app/src/main/java/org/eu/net/zendai/kanbankt/TaskDB.kt
package org.eu.net.zendai.kanbankt

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import java.util.*

@Database(entities = [Task::class], version = 1, exportSchema = false)
abstract class TaskDB : RoomDatabase() {
    abstract fun taskDao(): TaskDao

    private class TaskDatabaseCallback(
        private val scope: CoroutineScope
    ) : Callback() {
        override fun onCreate(db: SupportSQLiteDatabase) {
            super.onCreate(db)
            INSTANCE?.let { database ->
                scope.launch {
                    populateDatabase(database.taskDao())
                }
            }
        }
        suspend fun populateDatabase(taskDao: TaskDao) {
            taskDao.deleteAll()
            val currentTime = Calendar.getInstance().time.toString()
            val task = Task(0,
                "Welcome to KanbanKt!",
                "Create a new Task by clicking on the '+' in the bottom right corner of your screen ",
                "Not started",
                currentTime
            )
            taskDao.insert(task)
        }
    }

    companion object {
        @Volatile
        private var INSTANCE: TaskDB? = null
        fun getDatabase(
            context: Context,
            scope: CoroutineScope
        ): TaskDB {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    TaskDB::class.java,
                    "tasks_database"
                ).addCallback(TaskDatabaseCallback(scope)).build()
                INSTANCE = instance
                // Return instance
                instance
            }
        }
    }
}